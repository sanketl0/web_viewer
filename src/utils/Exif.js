/**
 * Read float value from EXIF tags (to handle fractions & all)
 * @param {*} val The input EXIF tag value
 * @returns {number|undefined} The parsed value, or undefined if value is not readable
 * @private
 */
export function getExifFloat(val) {
	// Null-like values
	if(
		[null, undefined, ""].includes(val)
		|| typeof val === "string" && val.trim() === ""
	) {
		return undefined;
	}
	// Already valid number
	else if(typeof val === "number") {
		return val;
	}
	// String
	else if(typeof val === "string") {
		// Check if looks like a fraction
		if(/^-?\d+(\.\d+)?\/-?\d+(\.\d+)?$/.test(val)) {
			const parts = val.split("/").map(p => parseFloat(p));
			return parts[0] / parts[1];
		}

		// Try a direct cast to float
		try { return parseFloat(val); }
		catch(e) {} // eslint-disable-line no-empty

		// Unrecognized
		return undefined;
	}
	else { return undefined; }
}

/**
 * Find in picture metadata the GPS precision.
 * @param {object} picture The GeoJSON picture feature
 * @returns {string} The precision value (poor, fair, moderate, good, excellent, ideal, unknown)
 * @private
 */
export function getGPSPrecision(picture) {
	let quality = "‚ùì";
	const gpsHPosError = picture?.properties?.["quality:horizontal_accuracy"] || getExifFloat(picture?.properties?.exif?.["Exif.GPSInfo.GPSHPositioningError"]);
	const gpsDop = getExifFloat(picture?.properties?.exif?.["Exif.GPSInfo.GPSDOP"]);
	
	if(gpsHPosError !== undefined) {
		quality = `${gpsHPosError} m`;
	}
	else if(gpsDop !== undefined) {
		if(gpsDop < 1) { quality = "ideal"; }
		else if(gpsDop < 2) { quality = "excellent"; }
		else if(gpsDop < 5) { quality = "good"; }
		else if(gpsDop < 10) { quality = "moderate"; }
		else if(gpsDop < 20) { quality = "fair"; }
		else { quality = "poor"; }
	}

	return quality;
}

/**
 * Compute PSV sphere correction based on picture metadata & EXIF tags.
 * @param {object} picture The GeoJSON picture feature
 * @returns {object} The PSV sphereCorrection value
 * @private
 */
export function getSphereCorrection(picture) {
	// Photo direction
	let dir = picture.properties?.["view:azimuth"];
	if(dir === undefined) {
		const v = getExifFloat(picture.properties?.exif?.["Exif.GPSInfo.GPSImgDirection"]);
		if(v !== undefined) {
			dir = v;
		}
	}
	dir = dir || 0;

	// Yaw
	let yaw = picture.properties?.["pers:yaw"];
	let exifFallbacks = ["Xmp.GPano.PoseHeadingDegrees", "Xmp.Camera.Yaw", "Exif.MpfInfo.MPFYawAngle"];
	if(yaw === undefined) {
		for(let exif of exifFallbacks) {
			const v = getExifFloat(picture.properties?.exif?.[exif]);
			if(v !== undefined) {
				yaw = v;
				break;
			}
		}
	}
	yaw = yaw || 0;

	// Check if yaw is applicable: different from photo direction
	if(Math.round(dir) === Math.round(yaw) && yaw > 0) {
		console.warn("Picture with UUID", picture.id, "has same GPS Image direction and Yaw, could cause rendering issues");
		// yaw = 0;
	}

	// Pitch
	let pitch = picture.properties?.["pers:pitch"];
	exifFallbacks = ["Xmp.GPano.PosePitchDegrees", "Xmp.Camera.Pitch", "Exif.MpfInfo.MPFPitchAngle"];
	if(pitch === undefined) {
		for(let exif of exifFallbacks) {
			const v = getExifFloat(picture.properties?.exif?.[exif]);
			if(v !== undefined) {
				pitch = v;
				break;
			}
		}
	}
	pitch = pitch || 0;

	// Roll
	let roll = picture.properties?.["pers:roll"];
	exifFallbacks = ["Xmp.GPano.PoseRollDegrees", "Xmp.Camera.Roll", "Exif.MpfInfo.MPFRollAngle"];
	if(roll === undefined) {
		for(let exif of exifFallbacks) {
			const v = getExifFloat(picture.properties?.exif?.[exif]);
			if(v !== undefined) {
				roll = v;
				break;
			}
		}
	}
	roll = roll || 0;

	// Send result
	return pitch !== 0 && roll !== 0 ? {
		pan:  yaw * Math.PI / 180,
		tilt: pitch * Math.PI / 180,
		roll: roll * Math.PI / 180,
	} : {};
}

/**
 * Compute PSV panoData for cropped panorama based on picture metadata & EXIF tags.
 * @param {object} picture The GeoJSON picture feature
 * @returns {object} The PSV panoData values
 * @private
 */
export function getCroppedPanoData(picture) {
	let res;

	if(picture.properties?.["pers:interior_orientation"]) {
		if(
			picture.properties["pers:interior_orientation"]?.["visible_area"]
			&& picture.properties["pers:interior_orientation"]?.["sensor_array_dimensions"]
		) {
			const va = picture.properties["pers:interior_orientation"]["visible_area"];
			const sad = picture.properties["pers:interior_orientation"]["sensor_array_dimensions"];
			try {
				res = {
					fullWidth: parseInt(sad[0]),
					fullHeight: parseInt(sad[1]),
					croppedX: parseInt(va[0]),
					croppedY: parseInt(va[1]),
					croppedWidth: parseInt(sad[0]) - parseInt(va[2]) - parseInt(va[0]),
					croppedHeight: parseInt(sad[1]) - parseInt(va[3]) - parseInt(va[1]),
				};
			}
			catch(e) {
				console.warn("Invalid pers:interior_orientation values for cropped panorama "+picture.id);
			}
		}
	}

	if(!res && picture.properties?.exif) {
		try {
			res = {
				fullWidth: parseInt(picture.properties.exif?.["Xmp.GPano.FullPanoWidthPixels"]),
				fullHeight: parseInt(picture.properties.exif?.["Xmp.GPano.FullPanoHeightPixels"]),
				croppedX: parseInt(picture.properties.exif?.["Xmp.GPano.CroppedAreaLeftPixels"]),
				croppedY: parseInt(picture.properties.exif?.["Xmp.GPano.CroppedAreaTopPixels"]),
				croppedWidth: parseInt(picture.properties.exif?.["Xmp.GPano.CroppedAreaImageWidthPixels"]),
				croppedHeight: parseInt(picture.properties.exif?.["Xmp.GPano.CroppedAreaImageHeightPixels"]),
			};
		}
		catch(e) {
			console.warn("Invalid XMP.GPano values for cropped panorama "+picture.id);
		}
	}

	// Check if crop is really necessary
	if(res) {
		res = Object.fromEntries(Object.entries(res || {}).filter(e => !isNaN(e[1])));
		if(res.fullWidth == res.croppedWidth && res.fullHeight == res.croppedHeight) {
			res = {};
		}
	}

	return res || {};
}