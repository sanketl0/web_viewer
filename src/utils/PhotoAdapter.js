import { EquirectangularTilesAdapter } from "@photo-sphere-viewer/equirectangular-tiles-adapter";

/**
 * Override of PSV EquirectangularTilesAdapter for fine-tweaking.
 * @private
 */
export default class PhotoAdapter extends EquirectangularTilesAdapter {
	constructor(viewer, config) {
		super(viewer, config);
		this._shouldGoFast = config.shouldGoFast || (() => true);
	}

	/**
	 * Override to skip loading SD images according to shouldGoFast option.
	 */
	loadTexture(panorama, loader) {
		if(!panorama.origBaseUrl) { panorama.origBaseUrl = panorama.baseUrl; }
		else { panorama.baseUrl = panorama.origBaseUrl; }
		
		// Fast mode + thumbnail available + no HD image loaded yet + flat picture
		if(
			this._shouldGoFast()
			&& panorama.thumbUrl
			&& !panorama.hdLoaded
			&& panorama.rows == 1
		) {
			panorama.baseUrl = panorama.thumbUrl;
		}

		return super.loadTexture(panorama, loader).then(data => {
			if(panorama.baseUrl === panorama.origBaseUrl) {
				panorama.hdLoaded = true;
			}
			return data;
		});
	}

	/**
	 * Override to skip loading tiles according to shouldGoFast option.
	 * @private
	 */
	__loadTiles(tiles) {
		if(!this._shouldGoFast()) { super.__loadTiles(tiles); }
	}
}
