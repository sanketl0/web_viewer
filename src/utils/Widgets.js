// Every single icon imported separately to reduce bundle size
import { icon } from "@fortawesome/fontawesome-svg-core";
import { faXmark } from "@fortawesome/free-solid-svg-icons/faXmark";
import { faCircleExclamation } from "@fortawesome/free-solid-svg-icons/faCircleExclamation";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";
import { faMagnifyingGlass } from "@fortawesome/free-solid-svg-icons/faMagnifyingGlass";
import { faCircleNotch } from "@fortawesome/free-solid-svg-icons/faCircleNotch";
import { faCheck } from "@fortawesome/free-solid-svg-icons/faCheck";
import { faCopy } from "@fortawesome/free-solid-svg-icons/faCopy";
import { faStar } from "@fortawesome/free-solid-svg-icons/faStar";
import { faStar as farStar } from "@fortawesome/free-regular-svg-icons/faStar";
import { QUALITYSCORE_VALUES } from "./Utils";


/**
 * Creates a new button, already styled
 * @param {string} id The component ID
 * @param {string|Element} content The text content
 * @param {string} [title] A title label on overlay
 * @param {string[]} [classes] List of CSS classes to add
 * @returns {Element} The created button
 * @private
 */
export function createButton(id, content = null, title = null, classes = []) {
	const btn = document.createElement("button");
	if(content) {
		if(content instanceof HTMLElement || content instanceof Node) {
			btn.appendChild(content);
		}
		else {
			btn.innerHTML = content;
		}
	}
	btn.id = id;
	if(Array.isArray(classes)) {
		classes = classes.filter(c => c != null && c.length > 0);
	}
	btn.classList.add("gvs-btn", "gvs-widget-bg", ...classes);
	if(title) { btn.title = title; }
	return btn;
}

/**
 * Creates a new "expandable" button, already styled
 * @param {string} id The component ID
 * @param {object} icon The FontAwesome icon definition
 * @param {string} label The label text
 * @param {Widgets} container The widgets container
 * @param {string[]} [classes] List of CSS classes to add
 * @returns {Element} The created button
 * @private
 */
export function createExpandableButton(id, icon, label, container, classes = []) {
	const btn = document.createElement("button");
	btn.id = id;
	btn.appendChild(fa(icon));
	if(!container._viewer.isWidthSmall()) {
		btn.appendChild(document.createTextNode(label));
		if(classes.includes("gvs-filter-unset-btn")) {
			const resetIcon = fa(faXmark, { classes: ["gvs-filter-unset-btn"]});
			resetIcon.style.display = "none";
			btn.appendChild(resetIcon);
			classes = classes.filter(v => v !== "gvs-filter-unset-btn");
		}
		btn.appendChild(fa(faChevronDown));
	}
	else {
		btn.title = label;
	}
	btn.classList.add("gvs-btn", "gvs-widget-bg", "gvs-btn-expandable", ...classes);
	btn.setActive = val => {
		let span = btn.querySelector(".gvs-filters-active");
		const resetIcon = btn.querySelector(".gvs-filter-unset-btn");
		const downIcon = btn.querySelector(".fa-chevron-down");
		if(val && !span) {
			span = document.createElement("span");
			span.classList.add("gvs-filters-active");
			const svg = btn.querySelector("svg");
			if(svg.nextSibling) { btn.insertBefore(span, svg.nextSibling); }
			else { btn.appendChild(span); }
			if(resetIcon) {
				resetIcon.style.display = null;
				downIcon.style.display = "none";
			}
		}
		else if(!val && span) {
			span.remove();
			if(resetIcon) {
				resetIcon.style.display = "none";
				downIcon.style.display = null;
			}
		}
	};
	return btn;
}

/**
 * Creates a new search bar
 * @param {string} id The bar ID
 * @param {string} placeholder The default label to display when search field is empty
 * @param {function} onInput Event handler for search text input (should return a Promise)
 * @param {function} onResultClick Event handler for result entry click
 * @param {Widgets} container The widgets container
 * @param {boolean} [nonClosingPanel] Should the search result closes other panels
 * @param {boolean} [reduced] Should the search bar be reduced by default ?
 * @param {Element} [preContent] DOM element to insert before search input
 * @returns {Element} The search bar
 * @private
 */
export function createSearchBar(
	id, placeholder, onInput, onResultClick,
	container, nonClosingPanel = false, reduced = false,
	preContent = null,
) {
	// Container
	const bar = document.createElement("div");
	bar.classList.add("gvs-widget-bg", "gvs-search-bar");
	bar.id = id;
	if(reduced) { bar.classList.add("gvs-search-bar-reducable"); }

	// Pre-content
	if(preContent) {
		bar.appendChild(preContent);
	}

	// Input field
	const input = document.createElement("input");
	input.type = "text";
	input.placeholder = placeholder;
	input.id = `${id}-input`;
	input.setAttribute("autocomplete", "off");
	bar.appendChild(input);
	const extendInput = () => {
		bar.classList.remove("gvs-search-bar-reduced");
	};
	const reduceInput = () => {
		bar.classList.add("gvs-search-bar-reduced");
	};
	if(reduced) { reduceInput(); }

	// Status icon
	const icon = document.createElement("span");
	icon.classList.add("gvs-search-bar-icon");
	const iconSearch = fa(faMagnifyingGlass);
	const iconLoading = fa(faCircleNotch, { classes: ["fa-spin"] });
	const iconEmpty = fa(faXmark);
	const iconWarn = fa(faCircleExclamation);
	icon.appendChild(iconSearch);
	bar.appendChild(icon);

	// List of results
	const list = createPanel(container, bar, [], ["gvs-search-bar-results"], nonClosingPanel);
	bar.appendChild(list);

	// Change status icon
	const switchIcon = newStatusIcon => {
		icon.innerHTML = "";
		icon.appendChild(newStatusIcon);
	};

	// Reset search bar
	const resetSearch = () => {
		if(bar._throttler) { clearTimeout(bar._throttler); }
		input.value = "";
		list.innerHTML = "";
		list._toggle(false);
		delete bar._lastSearch;
		switchIcon(iconSearch);
		onResultClick(null);
		if(reduced) { reduceInput(); }
	};
	bar.resetSearch = resetSearch;

	// Handle result item click
	input.goItem = (entry) => {
		if(reduced) {
			onResultClick(entry);
			resetSearch();
		}
		else {
			if(bar._throttler) { clearTimeout(bar._throttler); }
			input.value = entry.title;
			list.innerHTML = "";
			list._toggle(false);
			switchIcon(iconEmpty);
			onResultClick(entry);
		}
	};

	// Force item selection
	input.setItem = (text) => {
		if(bar._throttler) { clearTimeout(bar._throttler); }
		input.value = text;
		list.innerHTML = "";
		list._toggle(false);
		switchIcon(iconEmpty);
		if(reduced) { extendInput(); }
	};

	// Handle search
	const goSearch = () => {
		if(bar._throttler) { clearTimeout(bar._throttler); }

		if(input.value.length === 0) {
			list.innerHTML = "";
			list._toggle(false);
			return;
		}

		bar._throttler = setTimeout(() => {
			list.innerHTML = "";
			list._toggle(false);
			switchIcon(iconLoading);

			onInput(input.value).then(data => {
				switchIcon(iconEmpty);
				list._toggle(true);

				if(!data || data.length == 0) {
					list.innerHTML = `<div class="gvs-search-empty">${container._t.gvs.search_empty}</li>`;
					return;
				}
				else if(data === true) {
					list._toggle(false);
					return;
				}

				data.forEach(entry => {
					const listEntry = document.createElement("div");
					listEntry.classList.add("gvs-search-bar-result");
					listEntry.innerHTML = `${entry.title}<br /><small>${entry?.subtitle || ""}</small>`;
					list.appendChild(listEntry);
					listEntry.addEventListener("click", () => input.goItem(entry));
				});
			}).catch(e => {
				console.error(e);
				switchIcon(iconWarn);
			});
		}, 250);
	};

	input.addEventListener("change", goSearch);
	input.addEventListener("keypress", goSearch);
	input.addEventListener("paste", goSearch);
	input.addEventListener("input", goSearch);
	icon.addEventListener("click", () => {
		if(icon.firstChild == iconEmpty || icon.firstChild == iconWarn) {
			resetSearch();
		}
		if(reduced && icon.firstChild == iconSearch) {
			if(!bar.classList.contains("gvs-search-bar-reduced")) { reduceInput(); }
			else { extendInput(); }
		}
	});

	return bar;
}

/**
 * Creates a panel associated to a button
 * @param {Widgets} container The widgets container
 * @param {Element} btn The component to associate to
 * @param {Element[]} [elements] DOM elements to append into
 * @param {str[]} [classes] CSS classes to add
 * @param {boolean} [nonClosingPanel] Should this panel closes other when opened
 * @returns {Element} The created panel
 * @private
 */
export function createPanel(container, btn, elements = [], classes = [], nonClosingPanel = false) {
	const panel = document.createElement("div");
	panel.id = btn.id + "-panel";
	if(Array.isArray(classes)) {
		classes = classes.filter(c => c != null && c.length > 0);
	}
	panel.classList.add("gvs-panel", "gvs-widget-bg", "gvs-hidden", ...classes);
	for(let e of elements) {
		panel.appendChild(e);
	}

	const togglePanel = (e, visible) => {
		if(e) { e.stopPropagation(); }
		if(visible === true) { panel.classList.remove("gvs-hidden"); }
		else if(visible === false) { panel.classList.add("gvs-hidden"); }
		else {
			panel.classList.toggle("gvs-hidden");
			visible = !panel.classList.contains("gvs-hidden");
		}

		// Hide all other panels
		if(visible && !nonClosingPanel) {
			closeOtherPanels(panel, container._viewer.container);
		}
	};
	panel._toggle = v => togglePanel(null, v);

	if(btn.tagName == "BUTTON") {
		btn.addEventListener("click", togglePanel);
		btn.addEventListener("hover", togglePanel);
	}

	return panel;
}

/**
 * Makes all previously opened panels closed if clicked outside of one.
 * @param {Element} target The DOM element which has been clicked
 * @param {Element} container The viewer container
 * @private
 */
export function closeOtherPanels(target, container) {
	const isPanel = p => (
		p.classList.contains("gvs-panel")
		|| p.classList.contains("gvs-search-bar-result")
		|| p.classList.contains("gvs-search-empty")
		|| p.classList.contains("gvs-search-bar-reducable")
	);

	// Find nearest panel
	if(!isPanel(target) && target?.parentNode) {
		target = target.parentNode;
		while(target instanceof Element) {
			if(isPanel(target)) { break; }
			else { target = target.parentNode; }
		}
	}

	// Click outside of open panel = closing
	for(const p of container.getElementsByClassName("gvs-panel")) {
		if(p != target && !p.contains(target) && !p.classList.contains("gvs-hidden")) {
			p.classList.add("gvs-hidden");
		}
	}
	for(const p of container.getElementsByClassName("gvs-search-bar-reducable")) {
		if(p != target && !p.contains(target) && !p.classList.contains("gvs-search-bar-reduced")) {
			p.resetSearch();
		}
	}
}

/**
 * Creates a new group of elements, already styled
 * @param {str} id 
 * @param {str} position (format: component-corner, with component = main/mini, and corner = top-left, top-right, top, bottom-left, bottom, bottom-right)
 * @param {Element[]} [elements] The children elements to add
 * @param {str[]} [classes] The CSS classes to add
 * @returns {Element} The created group
 * @private
 */
export function createGroup(id, position, container, elements = [], classes = []) {
	const group = document.createElement("div");
	group.id = id;
	if(Array.isArray(classes)) {
		classes = classes.filter(c => c != null && c.length > 0);
	}
	group.classList.add("gvs-group", ...classes);
	for(let e of elements) {
		group.appendChild(e);
	}
	container._corners[position].appendChild(group);
	return group;
}

/**
 * Make all buttons with data-copy=* or data-input=* attributes able to copy to clipboard.
 * 
 * @param {Element} container The parent container
 * @param {object} t The translation container
 * @private
 */
export function enableCopyButton(container, t) {
	for(let btn of container.getElementsByTagName("button")) {
		const field = btn.getAttribute("data-input");
		const copy = btn.getAttribute("data-copy");
		if(field || copy) {
			btn.addEventListener("click", () => {
				let text;
				if(field) {
					const inputField = document.getElementById(field);
					text = inputField.innerText || inputField.value;
				}
				else if(copy) {
					text = btn.getAttribute("data-copy");
				}
				navigator.clipboard.writeText(text);
				const btnOrigContent = btn.innerHTML;
				btn.innerHTML = `${t.gvs.copied} ${fat(faCheck)}`;
				btn.classList.add("gvs-btn-active");
				setTimeout(() => {
					btn.innerHTML = btnOrigContent;
					btn.classList.remove("gvs-btn-active");
				}, 2000);
			});
		}
	}
}

/**
 * Make a button usable
 * @param {Element} btn
 * @private 
 */
export function enableButton(btn) {
	btn.removeAttribute("disabled");
}

/**
 * Make a button unusable
 * @param {Element} btn
 * @private
 */
export function disableButton(btn) {
	btn.setAttribute("disabled", "");
}

/**
 * Transform Font Awesome icon definition into HTML element
 * @param {IconDefinition} i The icon to use
 * @param {object} [o] [FontAwesome icon parameters](https://origin.fontawesome.com/docs/apis/javascript/methods#icon-icondefinition-params)
 * @returns {Element} HTML element
 * @private
 */
export function fa(i, o) {
	return icon(i, o).node[0];
}

/**
 * Transform Font Awesome icon definition into HTML text
 * @param {IconDefinition} i The icon to use
 * @param {object} [o] [FontAwesome icon parameters](https://origin.fontawesome.com/docs/apis/javascript/methods#icon-icondefinition-params)
 * @returns {string} HTML element as text
 * @private
 */
export function fat(i, o) {
	return icon(i, o).html[0];
}

/**
 * Table cell with a copy link
 * @private
 */
export function createLinkCell(id, url, title, buttonTitle) {
	const link = document.createElement("a");
	link.href = url;
	link.target = "_blank";
	link.title = title;
	link.textContent = id;

	const buttonContainer = createButtonSpan(`${fat(faCopy)} ${buttonTitle}`, id);
	return [link, buttonContainer];
}

/**
 * Create a light table
 * @private 
 */
export function createTable(className, rows) {
	const table = document.createElement("table");
	table.className = className;

	rows.forEach(({ section, value, values, classes }) => {
		const tr = document.createElement("tr");
		const th = document.createElement("th");
		th.scope = "row";
		th.textContent = section;
		tr.appendChild(th);

		const td = document.createElement("td");
		if(classes) { td.classList.add(...classes); }
		if(values) { values.forEach(v => td.appendChild(v)); }
		else if(value instanceof HTMLElement) { td.appendChild(value); }
		else { td.innerHTML = value; }
		tr.appendChild(td);

		table.appendChild(tr);
	});

	return table;
}

/**
 * Create block header
 * @private
 */
export function createHeader(tag, innerHTML) {
	const header = document.createElement(tag);
	header.innerHTML = innerHTML;
	return header;
}

/**
 * Create copy to clipboard button
 * @private
 */
export function createButtonSpan(innerHTML, dataCopy = null) {
	const button = document.createElement("button");
	button.innerHTML = innerHTML;
	if (dataCopy) button.setAttribute("data-copy", dataCopy);

	const span = document.createElement("span");
	span.className = "gvs-input-btn";
	span.appendChild(button);

	return span;
}

/**
 * Create an input label
 * @private
 */
export function createLabel(forAttr, text, faIcon = null) {
	const label = document.createElement("label");
	label.htmlFor = forAttr;
	if(faIcon) { label.appendChild(fa(faIcon)); }
	label.appendChild(document.createTextNode(text));
	return label;
}

/**
 * Show a grade in a nice, user-friendly way
 * @param {number} grade The obtained grade
 * @returns {string} Nice to display grade display
 * @private
 */
export function showGrade(grade, t) {
	let label = "<span class=\"gvs-grade\">";

	for(let i=1; i <= grade; i++) {
		label += fat(faStar);
	}
	for(let i=grade+1; i <= 5; i++) {
		label += fat(farStar);
	}
	
	label += "</span> (";
	if(grade === null) { label += t.gvs.metadata_quality_missing+")"; }
	else { label += grade + "/5)"; }
	return label;
}

/**
 * Displays a nice QualityScore
 * @param {number} grade The 1 to 5 grade
 * @returns {Element} The HTML code for showing the grade
 * @private
 */
export function showQualityScore(grade) {
	const span = document.createElement("span");

	for(let i=1; i <= QUALITYSCORE_VALUES.length; i++) {
		const pv = QUALITYSCORE_VALUES[i-1];
		const sub = document.createElement("span");
		sub.appendChild(document.createTextNode(pv.label));
		sub.classList.add("gvs-qualityscore");
		sub.style.backgroundColor = pv.color;
		if(i === (6-grade)) {
			sub.classList.add("gvs-qualityscore-selected");
		}
		span.appendChild(sub);
	}

	return span;
}