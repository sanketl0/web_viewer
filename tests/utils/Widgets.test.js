import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";
import * as Widgets from "../../src/utils/Widgets";

describe("createButton", () => {
	it("works", () => {
		const child = document.createElement("span");
		const btn = Widgets.createButton("mybtn", child, "blabla", ["gvs-blabla"]);
		expect(btn.className).toBe("gvs-btn gvs-widget-bg gvs-blabla");
		expect(btn.id).toBe("mybtn");
		expect(btn.children[0]).toBe(child);
		expect(btn.title).toBe("blabla");
	});

	it("works without options", () => {
		const child = document.createElement("span");
		const btn = Widgets.createButton("mybtn");
		expect(btn.className).toBe("gvs-btn gvs-widget-bg");
		expect(btn.id).toBe("mybtn");
		expect(btn.children.length).toBe(0);
		expect(btn.title).toBe("");
	});
});

describe("createExpandableButton", () => {
	it("works with large container", () => {
		const container = { _viewer: { isWidthSmall: () => false } };
		const btn = Widgets.createExpandableButton("mybtn", faChevronDown, "blabla", container, ["gvs-blabla"]);
		expect(btn.className).toBe("gvs-btn gvs-widget-bg gvs-btn-expandable gvs-blabla");
		expect(btn.innerHTML).toMatchSnapshot();
		expect(btn.id).toBe("mybtn");
	});

	it("works with small container", () => {
		const container = { _viewer: { isWidthSmall: () => true } };
		const btn = Widgets.createExpandableButton("mybtn", faChevronDown, "blabla", container, ["gvs-blabla"]);
		expect(btn.className).toBe("gvs-btn gvs-widget-bg gvs-btn-expandable gvs-blabla");
		expect(btn.innerHTML).toMatchSnapshot();
		expect(btn.id).toBe("mybtn");
	});
});

describe("createSearchBar", () => {
	it("works", () => {
		const container = { _t: {} };
		const btn = Widgets.createSearchBar("mysrch", "no res", jest.fn(), jest.fn(), container);
		expect(btn.className).toBe("gvs-widget-bg gvs-search-bar");
		expect(btn.innerHTML).toMatchSnapshot();
		expect(btn.id).toBe("mysrch");
	});
});

describe("createPanel", () => {
	it("works", () => {
		const btn = document.createElement("button");
		btn.id = "btngvs";
		btn.addEventListener = jest.fn();
		const child = document.createElement("span");
		const pnl = Widgets.createPanel({}, btn, [child], ["gvs-blabla"]);
		expect(pnl.id).toBe("btngvs-panel");
		expect(pnl.className).toBe("gvs-panel gvs-widget-bg gvs-hidden gvs-blabla");
		expect(pnl.innerHTML).toMatchSnapshot();
		expect(btn.addEventListener.mock.calls).toMatchSnapshot();
	});
});

describe("createGroup", () => {
	it("works", () => {
		const container = { _corners: { "main-top-left": document.createElement("div") } };
		const child = document.createElement("span");
		const grp = Widgets.createGroup("grp", "main-top-left", container, [child], ["gvs-blabla"]);
		expect(grp.className).toBe("gvs-group gvs-blabla");
		expect(grp.innerHTML).toMatchSnapshot();
		expect(grp.id).toBe("grp");
		expect(container._corners["main-top-left"].children[0]).toBe(grp);
	});
});

describe("enableButton", () => {
	it("works", () => {
		const btn = document.createElement("button");
		btn.setAttribute("disabled", "");
		Widgets.enableButton(btn);
		expect(btn.getAttribute("disabled")).toBeNull();
	});
});

describe("disableButton", () => {
	it("works", () => {
		const btn = document.createElement("button");
		Widgets.disableButton(btn);
		expect(btn.getAttribute("disabled")).toBe("");
	});
});

describe("fa", () => {
	it("works", () => {
		const res = Widgets.fa(faChevronDown);
		expect(res).toMatchSnapshot();
	});
});

describe("fat", () => {
	it("works", () => {
		const res = Widgets.fat(faChevronDown);
		expect(res).toMatchSnapshot();
	});
});