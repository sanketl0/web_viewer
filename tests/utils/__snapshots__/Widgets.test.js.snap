// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExpandableButton works with large container 1`] = `"<svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-down\\" class=\\"svg-inline--fa fa-chevron-down\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z\\"></path></svg>blabla<svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-down\\" class=\\"svg-inline--fa fa-chevron-down\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z\\"></path></svg>"`;

exports[`createExpandableButton works with small container 1`] = `"<svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-down\\" class=\\"svg-inline--fa fa-chevron-down\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z\\"></path></svg>"`;

exports[`createGroup works 1`] = `"<span></span>"`;

exports[`createPanel works 1`] = `"<span></span>"`;

exports[`createPanel works 2`] = `
Array [
  Array [
    "click",
    [Function],
  ],
  Array [
    "hover",
    [Function],
  ],
]
`;

exports[`createSearchBar works 1`] = `"<input type=\\"text\\" placeholder=\\"no res\\" id=\\"mysrch-input\\" autocomplete=\\"off\\"><span class=\\"gvs-search-bar-icon\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"magnifying-glass\\" class=\\"svg-inline--fa fa-magnifying-glass\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z\\"></path></svg></span><div id=\\"mysrch-panel\\" class=\\"gvs-panel gvs-widget-bg gvs-hidden gvs-search-bar-results\\"></div>"`;

exports[`fa works 1`] = `
<svg
  aria-hidden="true"
  class="svg-inline--fa fa-chevron-down"
  data-icon="chevron-down"
  data-prefix="fas"
  focusable="false"
  role="img"
  viewBox="0 0 512 512"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
    fill="currentColor"
  />
</svg>
`;

exports[`fat works 1`] = `"<svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-down\\" class=\\"svg-inline--fa fa-chevron-down\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z\\"></path></svg>"`;
